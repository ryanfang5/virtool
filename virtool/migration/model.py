from datetime import datetime

from sqlalchemy import Column, DateTime, Integer, String

from virtool.pg.base import Base


class SQLRevision(Base):
    """Describes an applied data revision."""

    __tablename__ = "revisions"

    id = Column(Integer, primary_key=True)
    revision = Column(String, unique=True)
    applied_at = Column(DateTime, default=datetime.utcnow)
