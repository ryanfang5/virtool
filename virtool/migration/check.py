async def check_data_revision_version(mongo: AsyncIOMotorDatabase):
    """
    Check if the required MongoDB revision has been applied.

    Log a fatal error and exit if the required revision
    has not been applied.

    :param mongo: the application database object
    """

    if not await mongo.migrations.find_one({"revision_id": REQUIRED_MONGODB_REVISION}):
        logger.fatal(
            "The required MongoDB revision has not been applied: %s.",
            REQUIRED_MONGODB_REVISION,
        )
        sys.exit(1)
