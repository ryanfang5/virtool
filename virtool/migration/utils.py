from pathlib import Path


def get_revisions_path() -> Path:
    """
    Get the hardcoded path to data revision files.

    :return: the revisions path

    """
    return Path(__file__).parent.parent.parent / "assets/revisions"
