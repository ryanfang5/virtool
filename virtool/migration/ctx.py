import asyncio
import sys
from dataclasses import dataclass
from logging import getLogger

from motor.motor_asyncio import (
    AsyncIOMotorClient,
    AsyncIOMotorClientSession,
    AsyncIOMotorDatabase,
)
from openfga_sdk import OpenFgaApi
from orjson import orjson
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession, create_async_engine

import virtool.mongo.connect
from virtool.api.custom_json import dump_string
from virtool.authorization.utils import connect_openfga
from virtool.config.cls import MigrationConfig

logger = getLogger("migration")


@dataclass
class MigrationContextMongo:
    client: AsyncIOMotorClient
    database: AsyncIOMotorDatabase


@dataclass
class RevisionContext:
    """
    A context for a specific revision upgrade or downgrade.


    """

    mongo: AsyncIOMotorClientSession
    openfga: OpenFgaApi
    pg: AsyncSession


@dataclass
class MigrationContext:
    mongo: MigrationContextMongo
    openfga: OpenFgaApi
    pg: AsyncEngine

    async def with_revision_context(self):
        async with AsyncSession(
            self.pg
        ) as pg, self.mongo.client.start_session() as mongo:
            yield RevisionContext(mongo, self.openfga, pg)


async def create_migration_context(config: MigrationConfig) -> MigrationContext:
    """
    Create a migration context that provides access to MongoDB, OpenFGA, and PostgreSQL.

    Connect to all data services and expose their clients in the returned context object.

    :param config: the migration configuration
    :return: the migration context

    """
    if not config.postgres_connection_string.startswith("postgresql+asyncpg://"):
        logger.critical("Invalid PostgreSQL connection string")
        sys.exit(1)

    logger.info("Connecting to PostgreSQL")

    try:
        pg = create_async_engine(
            config.postgres_connection_string,
            json_serializer=dump_string,
            json_deserializer=orjson.loads,
            pool_recycle=1800,
        )
    except ConnectionRefusedError:
        logger.critical("Could not connect to PostgreSQL: Connection refused")
        sys.exit(1)

    mongo_database, openfga = await asyncio.gather(
        virtool.mongo.connect.connect(
            config.mongodb_connection_string,
            config.mongodb_name,
            skip_revision_check=True,
        ),
        connect_openfga(
            config.openfga_host, config.openfga_scheme, config.openfga_store_name
        ),
    )

    return MigrationContext(
        mongo=MigrationContextMongo(
            client=mongo_database.mongo_client,
            database=AsyncIOMotorClient(
                config.mongodb_connection_string
            ).get_default_database(),
        ),
        openfga=openfga,
        pg=pg,
    )
