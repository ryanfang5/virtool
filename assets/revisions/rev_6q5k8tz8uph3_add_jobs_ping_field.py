"""
Add jobs ping field

Revision ID: 6q5k8tz8uph3
Date: 2022-10-07 20:14:53.735862

"""
from datetime import datetime

import arrow
from syrupy.matchers import path_type

from virtool.migration.ctx import RevisionContext

# Revision identifiers.
name = "Add jobs ping field"
created_at = arrow.get("2022-10-07 20:14:53.735862")
revision_id = "6q5k8tz8uph3"
required_alembic_revision = None


async def upgrade(ctx: RevisionContext):
    await ctx.mongo.database.jobs.update_many(
        {"ping": {"$exists": False}}, {"$set": {"ping": None}}, session=ctx.mongo.session
    )


async def test_upgrade(ctx, snapshot):
    await ctx.mongo.database.jobs.insert_many(
        [
            {"_id": "a", "ping": None},
            {"_id": "b"},
            {"_id": "c", "ping": {"pinged_at": arrow.utcnow().naive}},
            {"_id": "d"},
        ]
    )

    await upgrade(ctx)

    assert await ctx.mongo.database.jobs.find().to_list(None) == snapshot(
        matcher=path_type({".*pinged_at": (datetime,)}, regex=True)
    )
