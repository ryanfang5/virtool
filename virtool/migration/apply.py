from logging import getLogger
from pathlib import Path
from typing import Optional

import arrow
from alembic.util import load_python_file
from motor.motor_asyncio import AsyncIOMotorClient
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession

from virtool.config.cls import MigrationConfig
from virtool.migration.cls import AppliedRevision, Revision
from virtool.migration.ctx import MigrationContext, create_migration_context
from virtool.migration.model import SQLRevision
from virtool.migration.utils import get_revisions_path

logger = getLogger("migration")


async def apply_to(
    config: MigrationConfig,
    revision_id: str = "latest",
):
    """
    Apply revisions up to a given revision_id, ``to``.

    Providing ``'latest'`` as for ``to`` will apply all required revisions up to latest.

    Applied revisions are recorded in the target database so already applied revisions
    are not reapplied in subsequent migrations.

    :param config: the configuration values for migration
    :param revision_id: the revision_id to update to or 'latest'
    """
    logger.info("Applying revisions up to '%s'", revision_id)

    ctx = create_migration_context(config)

    last_applied_revision = await fetch_last_applied_revision(ctx.pg)

    for revision in _load_revisions(get_revisions_path()):
        if last_applied_revision is None or (
            revision.created_at > last_applied_revision.created_at
        ):
            await apply_one_revision(ctx, revision)

        if revision_id != "latest" and revision.id == revision_id:
            break


async def apply_one_revision(ctx: MigrationContext, revision):
    """
    Apply a single revision to Virtool data sources.

    :param ctx: the migration context
    :param revision: the revision to apply
    """
    logger.info("Applying revision '%s'", revision.id)

    async with ctx.with_revision_context() as revision_ctx:
        await revision.upgrade(revision_ctx)

        await revision_ctx.pg.add(
            SQLRevision(
                applied_at=arrow.utcnow().naive,
                revision=revision.id,
            )
        )


def _load_revisions(revisions_path: Path) -> list[Revision]:
    """
    Load revision modules and sort by creation date.

    :param revisions_path: the path to the revisions directory
    :returns: a list of revisions sorted by creation date
    """
    revisions = []

    for module_path in revisions_path.iterdir():
        if module_path.suffix == ".py":
            module = load_python_file(str(module_path.parent), str(module_path.name))

            revisions.append(
                Revision(
                    id=getattr(module, "revision_id"),
                    created_at=arrow.get(getattr(module, "created_at"))
                    .floor("second")
                    .naive,
                    name=getattr(module, "name"),
                    upgrade=getattr(module, "upgrade"),
                )
            )

    revisions.sort(key=lambda r: r.created_at)

    logger.info("Found %s revisions", len(revisions))

    return revisions


async def fetch_last_applied_revision(
    pg: AsyncEngine,
) -> Optional[AppliedRevision]:
    """
    Pull the most recent revision applied to the MongoDB database.
    """
    async with AsyncSession(pg) as session:
        result = await session.execute(
            select(SQLRevision).order_by(SQLRevision.applied_at)
        )

        revision = await result.first()

        return (
            AppliedRevision(
                id=revision.id,
                applied_at=arrow.get(revision.applied_at).floor("second").naive,
                name=revision.name,
                revision=revision.revision,
            )
            if revision
            else None
        )
