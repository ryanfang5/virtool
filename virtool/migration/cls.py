from dataclasses import dataclass
from datetime import datetime
from typing import Awaitable, Callable

from virtool.migration.ctx import MigrationContext


@dataclass
class Revision:
    """A Virtool data revision loaded from the filesystem."""

    id: int
    created_at: datetime
    name: str
    revision: str
    upgrade: Callable[[MigrationContext], Awaitable[None]]


@dataclass
class AppliedRevision:
    """A Virtool data revision that has been applied to the instance."""

    id: int
    applied_at: datetime
    name: str
    revision: str
