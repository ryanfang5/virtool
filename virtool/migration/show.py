import datetime
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from subprocess import call
from typing import Callable, Awaitable

import arrow
from alembic.util import load_python_file

from virtool.config.cls import MigrationConfig
from virtool.migration.apply import apply_one_revision, load_revisions
from virtool.migration.ctx import create_migration_context, MigrationContext
from virtool.migration.pg import fetch_last_applied_revision, list_applied_revision_ids


class RevisionSource(str, Enum):
    """The source of a revision."""

    ALEMBIC = "alembic"
    VIRTOOL = "virtool"


@dataclass
class GenericRevision:
    id: str
    created_at: datetime.datetime
    name: str
    source: RevisionSource
    depends_on: str
    upgrade: Callable[[MigrationContext], Awaitable[None]]


def get_alembic_path() -> Path:
    """Get the path to the alembic directory."""
    return Path(__file__).parent.parent.parent / "assets/alembic/versions"


def get_revision_create_date(path: Path) -> datetime.datetime:
    """Get the date the revision was created."""
    with open(path, "r") as f:
        for line in f:
            if line.startswith("Create Date:"):
                return arrow.get(line.replace("Create Date:", "").strip()).naive

    raise ValueError("Could not find create date")


def get_revision_name(path: Path) -> str:
    """Get the date the revision was created."""
    with open(path, "r") as f:
        for line in f:
            return line.replace('"""', "").strip()


def load_alembic_revisions() -> list[GenericRevision]:
    """Load all alembic revisions."""
    revisions = []

    for path in get_alembic_path().iterdir():
        if path.suffix == ".py":
            module = load_python_file(str(path.parent), str(path.name))

            revisions.append(
                GenericRevision(
                    id=getattr(module, "revision"),
                    created_at=get_revision_create_date(path),
                    name=get_revision_name(path),
                    source=RevisionSource.ALEMBIC,
                    depends_on=None,
                    upgrade=None,
                )
            )

    return revisions


async def apply(config: MigrationConfig, revision_id: str):

    all_revisions = sorted(
        [*load_alembic_revisions(), *load_virtool_revisions()],
        key=lambda r: r.created_at,
    )

    ctx = await create_migration_context(config)

    last_applied_revision = await fetch_last_applied_revision(ctx.pg)

    for revision in all_revisions:

        if revision.id in await list_applied_revision_ids(ctx.pg):
            continue

        if last_applied_revision is None or (
            revision.created_at > last_applied_revision.created_at
        ):

            if revision.source == RevisionSource.VIRTOOL:

                await apply_one_revision(ctx, revision)

            else:

                call(["alembic", "upgrade", revision.id])

        if revision_id != "latest" and revision.id == revision_id:
            break


def load_virtool_revisions() -> list[GenericRevision]:
    """
    Load all virtool revisions as generic revisions.

    This is used to interleave Virtool revisions with Alembic revisions.

    :return: a list of generic revisions
    """
    revisions = load_revisions(Path(__file__).parent.parent.parent / "assets/revisions")

    return [
        GenericRevision(
            id=revision.id,
            created_at=revision.created_at,
            name=revision.name,
            source=RevisionSource.VIRTOOL,
            depends_on=revision.depends_on,
            upgrade=revision.upgrade,
        )
        for revision in revisions
    ]


def show_revisions():
    """Show all available revisions."""
    all_revisions = sorted(
        [*load_alembic_revisions(), *load_virtool_revisions()],
        key=lambda r: r.created_at,
    )

    print(f"Found {len(all_revisions)} revisions\n")

    print("Date\t\t", "ID\t", "Source", "Name", sep="\t")

    for revision in all_revisions:
        print(
            arrow.get(revision.created_at).format("YYYY-MM-DD HH:mm:ss"),
            revision.id,
            revision.source.value,
            revision.name,
            sep="\t",
        )
