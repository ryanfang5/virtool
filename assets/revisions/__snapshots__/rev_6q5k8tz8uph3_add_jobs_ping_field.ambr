# name: test_upgrade
  <class 'list'> [
    <class 'dict'> {
      '_id': 'a',
      'ping': None,
    },
    <class 'dict'> {
      '_id': 'b',
      'ping': None,
    },
    <class 'dict'> {
      '_id': 'c',
      'ping': <class 'dict'> {
        'pinged_at': <class 'datetime'>,
      },
    },
    <class 'dict'> {
      '_id': 'd',
      'ping': None,
    },
  ]
---
